rust/src/commands/mod.rs

```diff
@@ -0,0 +1,279 @@
+    #[error("Error in underlying DeltaTable")]
+    DeltaTableError {
+        /// Raw internal DeltaTableError
+        #[from]
+        source: DeltaTableError,
+    },
+
+    /// Errors occurring inside the DeltaWriter modules
+    #[error("Error in underlying storage backend")]
+    DeltaWriterError {
+        /// Raw internal StorageError
+        #[from]
+        source: DeltaWriterError,
+    },
+
+    /// Error returned when errors occur in underlying storage instance
+    #[error("Error in underlying storage backend")]
+    StorageError {
+        /// Raw internal StorageError
+        #[from]
+        source: StorageError,
+    },
+
+    /// Error returned when errors occur in underlying storage instance
+    #[error("Error handling arrow data")]
+    ArrowError {
+        /// Raw internal StorageError
+        #[from]
+        source: ArrowError,
+    },
+
+    /// Error returned for errors internal to Datafusion
+    #[error("Error handling arrow data")]
+    DataFusionError {
+        /// Raw internal DataFusionError
+        #[from]
+        source: DataFusionError,
+    },
+}
+
+fn to_datafusion_err(e: impl std::error::Error) -> DataFusionError {
+    DataFusionError::Plan(e.to_string())
+}
+
+/// High level interface for executing commands against a DeltaTable
+pub struct DeltaCommands {
+    table: DeltaTable,
+}
+
+impl DeltaCommands {
+    /// load table from uri
+    pub async fn try_from_uri<T>(uri: T) -> DeltaCommandResult<Self>
+    where
+        T: Into<String>,
+    {
+        let table = get_table_from_uri_without_update(uri.into())?;
+        Ok(Self { table })
+    }
+
+    /// Get a reference to the underlying table
+    pub fn table(&self) -> &DeltaTable {
+        &self.table
+    }
+
+    async fn execute(
+        &mut self,
+        operation: DeltaOperation,
+        plan: Arc<dyn ExecutionPlan>,
+    ) -> DeltaCommandResult<()> {
+        let transaction = Arc::new(DeltaTransactionPlan::new(
+            self.table.table_uri.clone(),
+            plan,
+            operation,
+            None,
+        ));
+
+        let _ = collect(transaction).await?;
+        self.table.update().await?;
+
+        Ok(())
+    }
+
+    /// Create a new Delta table
+    pub async fn create(
+        &mut self,
+        metadata: DeltaTableMetaData,
+        mode: Option<SaveMode>,
+    ) -> DeltaCommandResult<()> {
+        let operation = DeltaOperation::Create {
+            mode: mode.clone().unwrap_or(SaveMode::Ignore),
+            metadata: metadata.clone(),
+            location: self.table.table_uri.clone(),
+            // TODO get the protocol from somewhere central
+            protocol: Protocol {
+                min_reader_version: 1,
+                min_writer_version: 2,
+            },
+        };
+        let plan = Arc::new(CreateCommand::try_new(
+            &self.table.table_uri,
+            operation.clone(),
+        )?);
+
+        self.execute(operation, plan).await
```

Comment:
perhaps wrapping create command in a transaction is an overkill here and we can just call `table.create()` here?

Reply:
On its own definitely. Main reason it exists, is that it made it much easier to use of a table needs to be created from within other commands, and to have a consistent way individual commands are implemented. Also I was unsure in cases where we create the table as part of e.g. a write command. calling `table.create()` would generate a version and then the write would create another version. in the current implementation a single version / commit is created if the table is created as part of a write command. Not sure how this is handled in the reference implementation, but since table metadata is derived from the written data, i felt it more natural to be a single commit. In that I thought moving the actions around consistently was worth the overhead of having a execution plan. I guess it all depends on how we want to handle wrapping things inside a transaction in general.

Reply:
I agree with you and think we should keep the `CreateCommand` plan. For writes to an non-existence table, I would prefer us to combine the create table actions and add actions into a single commit too.

For this particular call site, I was thinking we can avoid all the datafusion overhead and create the table directly since we are not writing extra data.